Recrutify/Recrutify.Services/Services/CandidateService.cs

```diff
@@ -46,10 +52,51 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
             return result;
         }
 
-        public Task UpsertFeedbackAsync(Guid id, Guid projectId, CreateFeedbackDTO feedbackDto)
+        public async Task<CandidateDTO> GetCandidateWithProjectAsync(Guid id, Guid projectId)
         {
-            var feedback = _mapper.Map<Feedback>(feedbackDto);
-            return _candidateRepository.UpsertFeedbackAsync(id, projectId, feedback);
+            var candidate = await _candidateRepository.GetAsync(id);
+            if (candidate == null)
+            {
+                throw new NotFoundException();
+            }
+
+            var currentProjectResult = candidate.ProjectResults?.FirstOrDefault(x => x.ProjectId == projectId);
+            candidate.ProjectResults = new List<ProjectResult> { currentProjectResult };
+            return _mapper.Map<CandidateDTO>(candidate);
+        }
+
+        public async Task UpsertFeedbackAsync(Guid id, Guid projectId, CreateFeedbackDTO feedbackDto)
+        {
+            var projectResultWithFeedback = await _candidateRepository.GetAsync(id);
+            if (projectResultWithFeedback == null)
+            {
+                throw new NotFoundException();
+            }
+
+            var currentFeedback = projectResultWithFeedback?.ProjectResults?.FirstOrDefault(x => x.ProjectId == projectId)
+                ?.Feedbacks?.FirstOrDefault(x => x.UserId == feedbackDto.UserId && x.Type == _mapper.Map<FeedbackType>(feedbackDto.Type));
+            if (currentFeedback == null)
+            {
+                var newFeedback = _mapper.Map<Feedback>(feedbackDto);
+                newFeedback.CreatedOn = DateTime.UtcNow;
+                await _candidateRepository.CreateFeedbackAsync(id, projectId, newFeedback);
+            }
+            else
+            {
+                var projectresult = projectResultWithFeedback.ProjectResults.FirstOrDefault(x => x.ProjectId == projectId);
```

Comment:
rename to projectResult, move it to line 75, and on line 76 write projectResult?.Feedbacks?.FirstOrDefault(x => x.UserId == feedbackDto.UserId && x.Type == _mapper.Map<FeedbackType>(feedbackDto.Type));

Now you do search by projectId twice - on 76, 86 lines, it's bad

Reply:
donedo